cgi-bin/DW/BlobStore.pm

```diff
@@ -0,0 +1,157 @@
+#!/usr/bin/perl
+#
+# DW::BlobStore
+#
+# Meta storage API for storing arbitrary blobs of content by key.
+#
+# Authors:
+#     Mark Smith <mark@dreamwidth.org>
+#
+# Copyright (c) 2016 by Dreamwidth Studios, LLC.
+#
+# This program is free software; you may redistribute it and/or modify it under
+# the same terms as Perl itself.  For a copy of the license, please reference
+# 'perldoc perlartistic' or 'perldoc perlgpl'.
+#
+
+package DW::BlobStore;
+
+use strict;
+use v5.10;
+use Log::Log4perl;
+my $log = Log::Log4perl->get_logger( __PACKAGE__ );
+
+use DW::BlobStore::LocalDisk;
+use DW::BlobStore::MogileFS;
+use DW::BlobStore::S3;
+
+my $blobstores;
+
+sub _get_blobstores {
+    # If we've already created one, simply return it.
+    return $blobstores if defined $blobstores;
+
+    # Site owners must configure one of these. Disk is configured by default
+    # to point to our run directory which will work for test environs.
+    if ( exists $LJ::BLOBSTORE{s3} ) {
+        $log->debug( 'Initializing S3 blobstore.' );
+        push @{$blobstores ||= []}, DW::BlobStore::S3->init( %{$LJ::BLOBSTORE{s3}} );
+    }
+    if ( exists $LJ::BLOBSTORE{localdisk} ) {
+        $log->logcroak( 'Must only define a single $LJ::BLOBSTORE entry.' )
+            if defined $blobstores;
+        $log->debug( 'Initializing localdisk blobstore.' );
+        push @{$blobstores ||= []}, DW::BlobStore::LocalDisk->init( %{$LJ::BLOBSTORE{localdisk}} );
```

Comment:
It would be a good idea to add an example %LJ::BLOBSTORE to the config.  I don't see the format of the hash documented anywhere.  It looks like it's supposed to be a hash with a single key, the value of which is a hashref of a format that can't be determined without parsing through the submodule's code.